KwcShop/Kwc/Shop/Cart/Checkout/Payment/Wirecard/ConfirmLink/Controller.php

```diff
@@ -21,4 +21,82 @@ public function jsonConfirmOrderAction()
+            throw new Kwf_Exception_Client('Set wirecard settings in config. (wirecard.url, wirecard.merchant.id, wirecard.auth.username, wirecard.auth.password)');
+        }
+
+        $client = new Zend_Http_Client($wirecardUrl);
+        $client->setHeaders(array(
+            'Content-Type' => 'application/json',
+            'Accept' => 'application/json'
+        ));
+        $client->setAuth($wirecardUsername, $wirecardPassword, \Zend_Http_Client::AUTH_BASIC);
+        $custom = $this->getParam('custom') ? rawurldecode($this->getParam('custom')) : null;
+        $data = KwcShop_Kwc_Shop_Cart_Checkout_Payment_Wirecard_LogModel::decodeCallback($custom);
+        $postData = array(
+            'payment' => array(
+                'merchant-account-id' => array(
+                    'value' => $wirecardMerchantId
+                ),
+                'account-holder' => array(
+                    'first-name' => $this->getParam('firstname'),
+                    'last-name' => $this->getParam('lastname'),
+                ),
+                'locale' => $this->getParam('language'),
+                'request-id' => $data['data']['orderId'],
+                'transaction-type' => 'purchase',
+                'payment-methods' => array(
+                    'payment-method' => array(
+                        array(
+                            'name' => 'creditcard'
+                        )
+                    )
+                ),
+                'three-d' => array(
+                    'attempt-three-d' => 'true'
+                ),
+                'requested-amount' => array(
+                    'value' => floatval($this->getParam('amount')),
+                    'currency' => $this->getParam('currency')
+                ),
+                'success-redirect-url' => $this->getParam('successRedirectUrl'),
+                'fail-redirect-url' => $this->getParam('failRedirectUrl'),
+                'cancel-redirect-url' => $this->getParam('cancelRedirectUrl'),
+                'notifications' => array(
+                    'format' => 'application/json',
+                    'notification' => array(
+                        array(
+                            'url' => $this->getParam('successRedirectUrl'),
+                            'transaction-state' => 'success'
+                        ),
+                        array(
+                            'url' => $this->getParam('failRedirectUrl'),
+                            'transaction-state' => 'failed'
+                        )
+                    )
+                )
+            )
+        );
+        $client->setRawData(json_encode($postData));
+        try {
+            $response = $client->request('POST');
+            if ($response->isSuccessful()) {
+                $body = json_decode($response->getBody(), true);
+                header("Location: {$body['payment-redirect-url']}"); // redirect to wirecard checkout
```

Comment:
wenn diese action per AJAX aufgerufen wird (was normalerweise bei allen json* der fall ist) dann kann ein redirect nicht funktionieren - wie klappt das dann hier?

Reply:
@nsams Die Action wird vom POST einer html-Form aufgerufen (zuerst hatte ich's per AJAX). Aber abgesehen davon, funktioniert es hier nur wegen dem exit, das danach kommt. Ohne exit geht's nicht, weil dann der Controller ganz normal {"success": true} zurückschickt.

Reply:
aber im fehlerfall wird schon ein JSON zurückgegeben?

Reply:
@nsams Ja, ein json wird schon zurückgegeben. Was mir aber noch aufgefallen ist, und was wirklich nicht ideal ist: Da der Controller nicht per AJAX aufgerufen wird, sondern über die action einer Form werden sämtliche Responses einfach auf dem weißen Bildschirm ausgegeben. Im Erfolgsfall wird man weitergeleitet und alles ist gut. Im Fehlerfall wird die Response, bzw. ja auch die Exceptions einfach schwarz auf weiß auf dem Bildschirm ausgegeben und nicht in Form einer schön gestylten Fehlermeldung. Evtl. kann man das noch optimieren.

Reply:
Ja, mach das noch bitte. Errorhandling gehört schon immer auch dazu...